File: algorithms/414. Third Maximum Number/solution.py
class Solution(object):
    def thirdMax(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        unique = set(nums)
        if len(unique) < 3:
            return max(unique)
        unique.remove(max(unique))
        unique.remove(max(unique))
        return max(unique)

# 26 / 26 test cases passed.
# Status: Accepted
# Runtime: 55 ms
# beats 41.04%

File: algorithms/414. Third Maximum Number/solution.cc
class Solution {
public:
    int thirdMax(vector<int>& nums) {
        set<int> top3;
        for ( int n : nums ) {
            top3.insert(n);
            if ( top3.size() > 3 ) {
                top3.erase(top3.begin());
            }
        }
        return top3.size() == 3 ? *top3.begin() : *top3.rbegin();
    }
};
